drivers/gpio/gpio_npcx.c

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (c) 2020 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_gpio
+
+#include <kernel.h>
+#include <device.h>
+#include <drivers/gpio.h>
+#include <soc.h>
+
+#include "gpio_utils.h"
+#include "soc_gpio.h"
+#include "soc_miwu.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_npcx, LOG_LEVEL_ERR);
+
+/* GPIO module instances declarations */
+static struct device *gpio_devs[];
+static int gpio_devs_count;
+
+/* Driver config */
+struct gpio_npcx_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	/* GPIO controller base address */
+	uint32_t base;
+	/* IO port */
+	int port;
+	/* Size of wui mapping array */
+	int wui_size;
+	/* Mapping table between gpio bits and wui */
+	struct npcx_wui wui_maps[];
+};
+
+/* Driver data */
+struct gpio_npcx_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct gpio_npcx_config *)(dev)->config)
+
+#define DRV_DATA(dev) \
+	((struct gpio_npcx_data *)(dev)->data)
+
+#define HAL_INSTANCE(dev) \
+	(struct gpio_reg_t *)(DRV_CONFIG(dev)->base)
+
+
+/* Soc specific GPIO functions */
+struct device *soc_get_gpio_dev(int port)
+{
+	if (port >= gpio_devs_count)
+		return NULL;
+
+	return gpio_devs[port];
+}
+
+/* GPIO api functions */
+static int gpio_npcx_config(struct device *dev,
+			     gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct gpio_reg_t *inst = HAL_INSTANCE(dev);
+	uint32_t mask = BIT(pin);
+
+	/* Don't support simultaneous in/out mode */
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
+		return -ENOTSUP;
+	}
+
+	/* Don't support "open source" mode */
+	if (((flags & GPIO_SINGLE_ENDED) != 0U) &&
```

Comment:
There's a strong push in Zephyr to use MISRA compliant practices, which mandates a strict essential type model: no implicit type conversions.  Boolean is a MISRA essential type; bitwise operators apply to and produce values that are essentially unsigned type.

This is rule 14.4 in https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html#main-rules: 

> The controlling expression of an if statement and the controlling expression of an iteration-statement shall have essentially Boolean type

The correct style for this is `(flags & FLAG) != 0` over `!!(flags & FLAG)` or `(flags & FLAG)` even if the language specifies they're all equivalent.

Many of the uses near this one are violations, but this one is not.

Zephyr isn't enforcing this currently, but it will happen someday.

Reply:
That you for the background, Peter! Let's go with `(flags & FLAG) != 0` style then in all cases in this PR so we have less to change in the future.

Reply:
Done.